src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+{
+    if handle == INVALID_HANDLE_VALUE {
+        return Ok(INVALID_HANDLE_VALUE);
+    }
+
+    unsafe {
+        let mut new_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let ok = kernel32::DuplicateHandle(*CURRENT_PROCESS_HANDLE as HANDLE, handle,
+                                           other_process, &mut new_handle,
+                                           0, winapi::FALSE, flags);
+        if ok == winapi::FALSE {
+            Err(WinError::last("DuplicateHandle"))
+        } else {
+            Ok(new_handle)
+        }
+    }
+}
+
+// duplicate a handle in the current process
+fn dup_handle(handle: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process(handle, unsafe { kernel32::GetCurrentProcess() })
+}
+
+// duplicate a handle to the target process
+fn dup_handle_to_process(handle: HANDLE, other_process: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process_with_flags(handle, other_process, winapi::DUPLICATE_SAME_ACCESS)
+}
+
+// duplicate a handle to the target process, closing the source handle
+fn move_handle_to_process(handle: HANDLE, other_process: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process_with_flags(handle, other_process, winapi::DUPLICATE_CLOSE_SOURCE | winapi::DUPLICATE_SAME_ACCESS)
+}
+
+// get the current process handle as a handle in another process
+#[allow(dead_code)]
+fn current_process_handle_to_process(other_process: HANDLE) -> Result<HANDLE,WinError> {
+    unsafe {
+        let mut new_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let ok = kernel32::DuplicateHandle(kernel32::GetCurrentProcess(), kernel32::GetCurrentProcess(),
+                                           other_process, &mut new_handle,
+                                           winapi::PROCESS_DUP_HANDLE, winapi::FALSE, 0);
+        if ok == winapi::FALSE {
+            Err(WinError::last("DuplicateHandle"))
+        } else {
+            Ok(new_handle)
+        }
+    }
+}
+
+fn create_overlapped() -> Box<winapi::OVERLAPPED> {
+    Box::new(winapi::OVERLAPPED {
+        Internal: 0,
+        InternalHigh: 0,
+        Offset: 0,
+        OffsetHigh: 0,
+        hEvent: 0 as HANDLE
+    })
+}
+
+fn reset_overlapped(ov: &mut winapi::OVERLAPPED) {
+    ov.Internal = 0;
+    ov.InternalHigh = 0;
+    ov.Offset = 0;
+    ov.OffsetHigh = 0;
+    ov.hEvent = 0 as HANDLE;
+}
+
+
+#[derive(Debug, PartialEq)]
+struct WinHandle {
+    h: HANDLE
+}
+
+unsafe impl Send for WinHandle { }
+unsafe impl Sync for WinHandle { }
+
+impl Drop for WinHandle {
+    fn drop(&mut self) {
+        self.close();
+    }
+}
+
+impl Default for WinHandle {
+    fn default() -> WinHandle {
+        WinHandle { h: INVALID_HANDLE_VALUE }
+    }
+}
+
+impl Deref for WinHandle {
+    type Target = HANDLE;
+
+    #[inline]
+    fn deref(&self) -> &HANDLE {
+        &self.h
+    }
+}
+
+impl WinHandle {
+    fn new(h: HANDLE) -> WinHandle {
+        WinHandle { h: h }
+    }
```

Comment:
Well, my own stance on this -- and I actually do feel strongly about it -- is that an `unsafe` block should never ever rely on any conditions being upheld outside of the `unsafe` block, beyond Rust's own invariants enforced by the compiler. This is the only way to make sure unsafety is actually isolated to the `unsafe` blocks, and doesn't spill out throughout the code base, fundamentally breaking Rust's promise that nothing you do in safe code can ever cause memory errors.

I am aware that this can be tricky; and I'm not sure all existing code in `ipc-channel` sticks to this principle -- but it's what we should strive for, I believe.

(To really enforce this, any types that can be used in unsafe ways would actually need to go into sub-modules, exporting only safe -- or explicitly unsafe -- accessor methods... I haven't pursued this in the existing back-ends yet, because of the major code churn involved; but for the new code, you might indeed want to consider it from the beginning...)

As for the "trickling down", I can't say for sure without seeing the final result -- but my feeling is that it should actually do exactly what it's supposed to do: i.e. require us to explicitly mark as `unsafe` all parts of the code that have to be absolutely correct in order to ensure we don't ever get unsound behaviour...